### PR TITLE
Improve MFA:Gauth: TOTP HTTP request logging

### PR DESCRIPTION
**Pull request by linosgian**

### CODE REVIEW

File: support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/util/QRUtils.java [hunk 1]
```diff
@@ -66,7 +69,13 @@ public static void generateQRCode(final OutputStream stream, final String key,
                 .filter(j -> byteMatrix.get(i, j))
                 .forEach(j -> graphics.fillRect(i, j, 1, 1)));
 
-        ImageIO.write(image, "png", stream);
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        ImageIO.write(image, "PNG", out);
+        byte[] bytes = out.toByteArray();
+
+        String base64bytes = EncodingUtils.encodeBase64(bytes);
+        String src = "data:image/png;base64," + base64bytes;
```
Comment (mmoayyed) [inline support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/util/QRUtils.java:77]: Remove `"data:image/png;base64,"` and move it over to the actual template to be consistent with other areas in CAS doing the same.

File: support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/web/flow/OneTimeTokenAccountCheckRegistrationAction.java [hunk 2]
```diff
@@ -45,9 +48,12 @@ protected Event doExecute(final RequestContext requestContext) {
         if (acct == null || StringUtils.isBlank(acct.getSecretKey())) {
             val keyAccount = this.repository.create(uid);
             val keyUri = "otpauth://totp/" + this.label + ':' + uid + "?secret=" + keyAccount.getSecretKey() + "&issuer=" + this.issuer;
+            ByteArrayOutputStream QRcodeBase64 = new ByteArrayOutputStream();
+            QRUtils.generateQRCode(QRcodeBase64, keyUri, QRUtils.WIDTH_LARGE, QRUtils.WIDTH_LARGE);
+
             val flowScope = requestContext.getFlowScope();
             flowScope.put(FLOW_SCOPE_ATTR_ACCOUNT, keyAccount);
-            flowScope.put(FLOW_SCOPE_ATTR_ACCOUNT_URI, keyUri);
+            flowScope.put(FLOW_SCOPE_ATTR_QR_IMAGE_BASE64, QRcodeBase64);
```
Comment (mmoayyed) [inline support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/web/flow/OneTimeTokenAccountCheckRegistrationAction.java:56]: Renaming a flow-scope variable will surely break tests. Please review and adjust.

**Review by mmoayyed — CHANGES_REQUESTED** [review]
In principal, this looks good. 

- Please fix all build failures.
- Be sure to add/update tests
- Make sure you post the same PR for the master branch.

File: support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/util/QRUtils.java (hunk 1 above)
Comment (linosgian) [reply support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/util/QRUtils.java]: I've pushed a fixup to address this one

File: support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/web/flow/OneTimeTokenAccountCheckRegistrationAction.java (hunk 2 above)
Comment (linosgian) [reply support/cas-server-support-otp-mfa-core/src/main/java/org/apereo/cas/otp/web/flow/OneTimeTokenAccountCheckRegistrationAction.java:56]: Right, I've fixed the tests as well. We should be okay now

**Comment by linosgian** [comment]
I just got the time to review what you've said, excuse my late response
> In principal, this looks good.
> 
>     * Please fix all build failures.
There should be none now
> 
>     * Be sure to add/update tests
> 
I've updated the tests, where needed, but I am not sure which part would need more testing, meaning that the QR generation class already existed and it's quite hard to test that the produced QR code is in fact a valid QR code, and it also contains the values we need it to.
>     * Make sure you post the same PR for the master branch.

I'll open another PR for master shortly. Since this is a security fix, should we also backport it for 6.1.X? I am not sure how we could go about doing this, since my other PR for 6.1.X got automatically closed due to 6.1.X being on SPM mode.

**Comment by mmoayyed** [comment]
Thanks much for the changes. Will review.

> I'll open another PR for master shortly. Since this is a security fix,

I don't really see this as a "security fix". Security issues are expected to be reported as such:
https://apereo.github.io/cas/developer/Sec-Vuln-Response.html

Note the big yellow box.

So no backports.

**Comment by linosgian** [comment]
> Thanks much for the changes. Will review.
> 
> > I'll open another PR for master shortly. Since this is a security fix,
> 
> I don't really see this as a "security fix". Security issues are expected to be reported as such:
> https://apereo.github.io/cas/developer/Sec-Vuln-Response.html
> 
Right, that's a misstep on my part, but I still consider this one security issue that deserves a notification to the users,
since valid TOTP seeds can still be on their servers long after the fix is deployed. 
> Note the big yellow box.
> 
> So no backports.
> 
> > **[CAS - Vulnerability Response](https://apereo.github.io/cas/developer/Sec-Vuln-Response.html)**CAS - Enterprise Single Sign-On for the Web

**Comment by mmoayyed** [comment]
Since the information and details are already public, I am not sure how a notification can be handled without causing panic, specially when taking into account the fact that no one I know on the dev team has available cycles to cut a patch for 6.1.x. I'll have to discuss that with the sec team to see what might be the best approach, and whether anyone has any availability or interest to actually merge and accept a patch for 6.1. If not, it's m00t as our normal process cannot apply. 

I do appreciate your time and contributions; truly. In the future if you find something deemed as a security issue severe enough to merit an actual fix, please share that information privately so the process can be followed. Look for the appropriate process, and if you don't find one, please ask before stepping into action.

I'll send a note to the security team to see what happens.

PS You don't have to do a PR for master; you're welcome if you wanted to spare the time but this should be easily cherry-pickable.

**Comment by mmoayyed** [comment]
Message sent. Stand by please.

**Comment by linosgian** [comment]
> Since the information and details are already public, I am not sure how a notification can be handled without causing panic, specially when taking into account the fact that no one I know on the dev team has available cycles to cut a patch for 6.1.x. I'll have to discuss that with the sec team to see what might be the best approach, and whether anyone has any availability or interest to actually merge and accept a patch for 6.1. If not, it's m00t as our normal process cannot apply.
> 
> I do appreciate your time and contributions; truly. In the future if you find something deemed as a security issue severe enough to merit an actual fix, please share that information privately so the process can be followed. Look for the appropriate process, and if you don't find one, please ask before stepping into action.
> 
> I'll send a note to the security team to see what happens.
> 
> PS You don't have to do a PR for master; you're welcome if you wanted to spare the time but this should be easily cherry-pickable.

You're absolutely right about this one, I rushed into opening a PR for this one (this would probably be simple issue on Github if you had them open, with the same repercussions of course). I'll keep it in mind, in case I stumble upon something else. Regardless of whether you opt to backporting this to 6.1, I think that after pushing the fix, a notification of sorts should be sent out to users (through the mailing list?), urging for updates and further action on their side, e.g. wiping logs and/or forcibly resetting all users' TOTP keys.

**Comment by mmoayyed** [comment]
In discussing this with relevant teams, we have decided to proceed with a release this week with follow-up announcements where needed. @leleuj might be in touch to review and handle the release later in the week. Given the body the patch, the PR history/convos have been "sanitized" to make sure we can follow the usual process as much as it can be done. We will handle the port to the main/master branch as well.

As always, thank you for your patience and contribution.

**Comment by linosgian** [comment]
Alright, I'll be available in case more changes are needed here. The 2 failed builds might contain some flaky test, possibly? I've ran them locally and they seem to pass, and I cannot get a pointer as to what I could do to fix them. All in all, they seem unrelated to my changes

**Pull request merged**